// aws-jwt-verify/dist/cjs/error.js
"use strict";
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
exports.AssertionError = exports.NonRetryableFetchError = exports.FetchError = exports.JwksNotAvailableInCacheError = exports.WaitPeriodNotYetEndedJwkError = exports.KidNotFoundInJwksError = exports.JwtWithoutValidKidError = exports.JwkValidationError = exports.JwksValidationError = exports.Asn1DecodingError = exports.JwtInvalidClaimError = exports.ParameterValidationError = exports.JwtNotBeforeError = exports.JwtExpiredError = exports.JwtInvalidSignatureError = exports.JwtParseError = exports.JwtBaseError = void 0;
class JwtBaseError extends Error {
}
exports.JwtBaseError = JwtBaseError;
/**
 * JWT errors
 */
class JwtParseError extends JwtBaseError {
    constructor(msg, error) {
        const message = error != null ? `${msg}: ${error}` : msg;
        super(message);
    }
}
exports.JwtParseError = JwtParseError;
class JwtInvalidSignatureError extends JwtBaseError {
}
exports.JwtInvalidSignatureError = JwtInvalidSignatureError;
class JwtExpiredError extends JwtBaseError {
}
exports.JwtExpiredError = JwtExpiredError;
class JwtNotBeforeError extends JwtBaseError {
}
exports.JwtNotBeforeError = JwtNotBeforeError;
class ParameterValidationError extends JwtBaseError {
}
exports.ParameterValidationError = ParameterValidationError;
class JwtInvalidClaimError extends JwtBaseError {
}
exports.JwtInvalidClaimError = JwtInvalidClaimError;
/**
 * ASN.1 errors
 */
class Asn1DecodingError extends JwtBaseError {
}
exports.Asn1DecodingError = Asn1DecodingError;
/**
 * JWK errors
 */
class JwksValidationError extends JwtBaseError {
}
exports.JwksValidationError = JwksValidationError;
class JwkValidationError extends JwtBaseError {
}
exports.JwkValidationError = JwkValidationError;
class JwtWithoutValidKidError extends JwtBaseError {
}
exports.JwtWithoutValidKidError = JwtWithoutValidKidError;
class KidNotFoundInJwksError extends JwtBaseError {
}
exports.KidNotFoundInJwksError = KidNotFoundInJwksError;
class WaitPeriodNotYetEndedJwkError extends JwtBaseError {
}
exports.WaitPeriodNotYetEndedJwkError = WaitPeriodNotYetEndedJwkError;
class JwksNotAvailableInCacheError extends JwtBaseError {
}
exports.JwksNotAvailableInCacheError = JwksNotAvailableInCacheError;
/**
 * HTTPS fetch errors
 */
class FetchError extends JwtBaseError {
    constructor(uri, msg) {
        super(`Failed to fetch ${uri}: ${msg}`);
    }
}
exports.FetchError = FetchError;
class NonRetryableFetchError extends FetchError {
}
exports.NonRetryableFetchError = NonRetryableFetchError;
/**
 * Assertion errors
 */
class AssertionError extends JwtBaseError {
}
exports.AssertionError = AssertionError;
